### PR TITLE
monitor disk temperatures

### PR DESCRIPTION
**Pull request by paulfantom**
Add a module to monitor disk temperatures using running instance of daemonized hddtemp.

Script automatically connects to hddtemp running on localhost:7634, pulls disks data and presents them on one graph.

### CODE REVIEW

**Comment by ktsaou** [comment]
nice! thank you!
Can you post sample output of the `nc` command with data from hddtemp?
I would like to find a way to avoid the 3 forks in `hddtemp_update()`, so let me try...

**Comment by paulfantom** [comment]
example with 4 drives, with /dev/sda in sleep mode:
|/dev/sda|TOSHIBA DT01ACA300|SLP|*||/dev/sdb|WDC WD10JFCX-68N6GN0|36|C||/dev/sdc|KINGSTON SV300S37A60G|33|C||/dev/sdd|WDC WD10JFCX-68N6GN0|37|C|

**Comment by ktsaou** [comment]
ok, I think I did it without forks (I hope you don't mind I tried to optimize your code).

```sh
hddtemp_host="localhost"
hddtemp_port="7634"

declare -A hddtemp_disks=()

# disable shell expansion of filenames
# we need this because hddtemp returns *
set -f

hddtemp_get() {
        local x disks i=0 d t

        IFS="|" x=( $(nc $hddtemp_host $hddtemp_port) );

        disks=$(( ${#x[@]} / 5 ))

        while [ $i -lt $disks ]
        do
                # disk name
                d=${x[$((i * 5 + 1))]};

                # strip /dev/ from disk name
                d=${d/\/dev\//}

                # temperature - this will turn SLP to zero
                t=$(( ${x[$((i * 5 + 3))]} ))

                # store it
                hddtemp_disks[$d]=$t

                # examine the next disk
                i=$(( i + 1 ))
        done

        # if we did not get any values
        # return 1
        [ $disks -eq 0 ] && return 1

        # all is good
        return 0
}

# run it
hddtemp_get

# let's see what it did
declare -p hddtemp_disks

echo "number of disks: ${#hddtemp_disks[@]}"
echo "disk names: ${!hddtemp_disks[@]}"

# loop through all disks
for x in ${!hddtemp_disks[@]}
do
        echo "disk $x temperature: ${hddtemp_disks[$x]}"
done

```

Mine prints this (with your data):

```sh
declare -A hddtemp_disks='([sdd]="37" [sdc]="33" [sdb]="36" [sda]="0" )'
```

So, just call `hddtemp_get`. If it returns 1, it failed. If it returns 0, it found disks.
The example loop at the end is how to use it to generate charts and give values to netdata.

If you adapt your code to use this, only one `nc` will be executed per iteration and just one fork is make (for the `nc` command).

**Comment by paulfantom** [comment]
I've added your code to this module and tested it. If I run it in bash 4.3.3 on debian 8.5 it works as you say, but when i ran it with 'charts.d.plugin' i got:
declare -a hddtemp_disks='([0]="37" [1]="sdb" [2]="sdc" [3]="sdd")'

Also I have changed my code use less forks, from 4 to 2 (nc and awk).

**Comment by ktsaou** [comment]
ok. I'll merge yours.
Do you know if the awk you need is gawk or any other one?

**Comment by paulfantom** [comment]
I just got it working without awk and with some of your ideas.

**Comment by ktsaou** [comment]
ok. one last thing: in my tests with your file, I needed to run `set -f` before running the `nc` command and `set +f` after it to reset filename globbing. Without this strategy, an `*` in the hddtemp output was expanded to filenames in the current directory.

(I edited my script above to reflect this)

**Comment by ktsaou** [comment]
not "your file", but "your test data"

**Comment by paulfantom** [comment]
I saw that, for me it didn't change anything when executing with `charts.d.plugin`. However I will add `set -f` and `set +f` around `nc` command.

**Comment by paulfantom** [comment]
Also IFS needs to be resetted after setting it to '|'. Permanently changing IFS breaks other plugins, ex. nginx or php-fpm

**Comment by ktsaou** [comment]
> Also IFS needs to be resetted after setting it to '|'. Permanently changing IFS breaks other plugins, ex. nginx or php-fpm

I think this is not needed. When you set a variable in the same line with a command, the variable is just set for that command.

Example:

```sh
# declare -p IFS
declare -- IFS="
"

# IFS="|" echo hello
hello

# declare -p IFS
declare -- IFS="
"
```

It did not change...

**Comment by ktsaou** [comment]
but it does no harm to leave it there...

**Comment by ktsaou** [comment]
No, you are right. It changes it.
Sorry!

**Comment by paulfantom** [comment]
No problem. Have you found any other problems with this plugin?

**Comment by ktsaou** [comment]
The only issue I see is that `hddtemp_update()` never returns 1. It should return 1 if the nc command fails, or the log will never stop complaining...

**Comment by ktsaou** [comment]
I need to leave. I'll merge it tomorrow.
Just post here when you are ready.

Many thanks! You are the best!

**Comment by paulfantom** [comment]
Already done.

**Comment by ktsaou** [comment]
merged. thanks!

File: charts.d/hddtemp.chart.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+	echo "CHART hddtemp.temperature 'disks_temp' 'temperature' 'Celsius' 'Disks temperature' 'hddtemp.temp' line $((hddtemp_priority)) $hddtemp_update_every"
+	for i in `seq 0 $((${#hddtemp_disks[@]}-1))`; do
+#		echo "DIMENSION ${hddtemp_disks[i]} ${disk_names[i]} absolute 1 1"
+		echo "DIMENSION ${hddtemp_disks[$i]} '' absolute 1 1"
+	done
+	return 0
+}
+
+# _update is called continiously, to collect the values
```
Comment (glensc) [inline charts.d/hddtemp.chart.sh]: continiously -> continuously

File: charts.d/hddtemp.chart.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+hddtemp_last=0
+hddtemp_count=0
+hddtemp_update() {
+#        local all=( `nc $hddtemp_host $hddtemp_port | sed -e 's/||/\n/g;s/^|//' | cut -d '|' -f3` )
+#	local all=( `nc $hddtemp_host $hddtemp_port | awk 'BEGIN { FS="|" };{i=4; while (i <= NF) {print $i+0;i+=5;};}'` )
+	OLD_IFS=$IFS
```
Comment (glensc) [inline charts.d/hddtemp.chart.sh]: OLD_IFS should be declared local

**Pull request merged**